### ISSUE TITLE
[Feature Request] Efficient model loading

### ISSUE THREAD

**Issue by pauldog** [microsoft/onnxruntime#15080]
### Describe the feature request

If you currently load a model of say 5GB, it will first load the model into ram taking 5GB, then it will do some sort of duplication, using another 5GB RAM. Spiking at 10GB RAM. It then transfers 5GB to the GPU and removes the 10GB from the RAM.

This is extremely wasteful and unnecessary. Because of short spike as observed:

![screenshotx](https://user-images.githubusercontent.com/33497043/225597224-5309ca27-b5e1-484e-8245-f9b380a4dd5d.png)

It means you need double the RAM you should actually require to run certain models.

I'm sure this can easily be overcome by loading the model piecemeal into RAM instead of inefficiently loading the whole model into RAM at once doing some masterful duplication and then deleting the entire thing.

Alternatively some of that work could be shifted to the VRAM.

Either way, this spike in RAM is just a symptom of very inefficient model loading.

Basically, the model loading could be done more efficiently to avoid this spike in RAM.

### Describe scenario use case

To load large models without having to buy 2x the RAM you actually should require.

**Comment by skottmckay** [comment]
It's not easily overcome with the current implementation.

ORT loads the model from a protobuf format .onnx file. It's up to protobuf to decide what arenas etc. are used to load that into memory. 

* The bulk of the memory is the initializer/weights data which may be stored in a packed format. 
  * i.e. we can't treat it as a simple array of the data type so it needs to be in a different format to actually run the model. 
  * a copy is required to do that, which leads to 2x the memory usage for each initializer
* We try and free individual things as we go along where possible, but that may not make a difference given protobuf controls what memory arenas etc. are used and a chunk of memory it owns cannot be freed until everything using that chunk is no longer in use.

Theoretically if you saved the model in some different format that when loaded you could directly use it for the initializer data you could avoid this copy, however a _lot_ of the code is written to operate on the protobuf types, such as all the optimizers, and rewriting that would be a significant undertaking.

**Comment by pauldog** [comment]
> Theoretically if you saved the model in some different format that when loaded you could directly use it for the initializer data you could avoid this copy, however a _lot_ of the code is written to operate on the protobuf types, such as all the optimizers, and rewriting **that would be a significant undertaking**.

Microsoft is a multi-billion dollar company with very intelligent people. 

What I'm hearing is "we agree Onnxruntime is inefficient and we don't want to fix it." Is that really the Microsoft attitude?

**Comment by faxu** [comment]
Hi @pauldog, please take a look at our [Code of Conduct](https://opensource.microsoft.com/codeofconduct/), which outlines our expectations for Microsoft open source community engagement.

We do our best to monitor and support community product feedback, and we expect community members to use respectful language when discussing issues.

**Comment by pranavsharma** [comment]
@pauldog If memory is a concern you can solve it the following way.
1. Create a session with the model by setting the optimized_file_path to serialize the optimized model file.
2. Externalize all weights from the optimized model.
3. Create OrtValues for each of the weights.
4. Feed them to ORT using [this API](https://github.com/microsoft/onnxruntime/blob/c6074f3a4b09cb44b46ca4422717deb067015bfa/include/onnxruntime/core/session/onnxruntime_c_api.h#L2440-L2441). Even though this API was originally developed to share weights between multiple models (sessions), it can still be used with a single session. It'll ensure the weights are allocated **only once** (by the user). Here's a [test](https://github.com/microsoft/onnxruntime/blob/59dfcfdce712a5d527c7b986fc41a8e4e14c702a/onnxruntime/test/shared_lib/test_inference.cc#L2261) that shows its usage.

**Comment by pauldog** [comment]
> @pauldog If memory is a concern you can solve it the following way.
> 
> 1. Create a session with the model by setting the optimized_file_path to serialize the optimized model file.
> 2. Externalize all weights from the optimized model.
> 3. Create OrtValues for each of the weights.
> 4. Feed them to ORT using [this API](https://github.com/microsoft/onnxruntime/blob/c6074f3a4b09cb44b46ca4422717deb067015bfa/include/onnxruntime/core/session/onnxruntime_c_api.h#L2440-L2441). Even though this API was originally developed to share weights between multiple models (sessions), it can still be used with a single session. It'll ensure the weights are allocated **only once** (by the user). Here's a [test](https://github.com/microsoft/onnxruntime/blob/59dfcfdce712a5d527c7b986fc41a8e4e14c702a/onnxruntime/test/shared_lib/test_inference.cc#L2261) that shows its usage.

Thanks. I'll give it a go!😃
Sorry if my previous words offended. 😔

**Comment by pauldog** [comment]
> @pauldog If memory is a concern you can solve it the following way.
> 
> 1. Create a session with the model by setting the optimized_file_path to serialize the optimized model file.
> 2. Externalize all weights from the optimized model.
> 3. Create OrtValues for each of the weights.
> 4. Feed them to ORT using [this API](https://github.com/microsoft/onnxruntime/blob/c6074f3a4b09cb44b46ca4422717deb067015bfa/include/onnxruntime/core/session/onnxruntime_c_api.h#L2440-L2441). Even though this API was originally developed to share weights between multiple models (sessions), it can still be used with a single session. It'll ensure the weights are allocated **only once** (by the user). Here's a [test](https://github.com/microsoft/onnxruntime/blob/59dfcfdce712a5d527c7b986fc41a8e4e14c702a/onnxruntime/test/shared_lib/test_inference.cc#L2261) that shows its usage.

Hi can you clarify what you mean by "externalize all weights"? I'm not sure I understand thanks.
